### PR TITLE
ui: Support wallets loaded dynamically

### PR DESCRIPTION
**Pull request by promag**
Add support in the UI for wallets loaded dynamically.

### CODE REVIEW

**Comment by promag** [comment]
@ryanofsky had to convert to shared pointer because of `QueuedConnection`. Do you have an alternative? Like instancing WalletModel on RPC thread and then make that the slot argument?

**Comment by jnewbery** [comment]
> @jnewbery mind testing this?

Will do! (next week)

**Comment by promag** [comment]
In order to test, suppose there are wallets `w1` and `w2`:
 - launch `bitcoin-qt -regtest -nowallet`
 - open console and run `loadwallet w1` (should should wallet)
 - run `loadwallet w2` (should show multi-wallet selector)

File: src/qt/bitcoin.cpp [hunk 1]
```diff
@@ -250,7 +251,8 @@ public Q_SLOTS:
     QTimer *pollShutdownTimer;
 #ifdef ENABLE_WALLET
     PaymentServer* paymentServer;
-    std::vector<WalletModel*> m_wallet_models;
+    std::list<WalletModel*> m_wallet_models;
```
Comment (promag) [inline src/qt/bitcoin.cpp]: Revert to vector.

File: src/qt/bitcoin.cpp [hunk 2]
```diff
@@ -447,6 +449,21 @@ void BitcoinApplication::requestShutdown()
...
 
+void BitcoinApplication::addWallet(WalletModel* walletModel)
+{
+#ifdef ENABLE_WALLET
+    window->addWallet(walletModel);
+    if (m_wallet_models.empty()) {
+        window->setCurrentWallet(walletModel->getWalletName());
+    }
+
+    connect(walletModel, SIGNAL(coinsSent(WalletModel*,SendCoinsRecipient,QByteArray)),
```
Comment (promag) [inline src/qt/bitcoin.cpp]: Improve format.

**Comment by ken2812221** [comment]
If the wallet folder name contains Chinese, it will create a new folder that has same wallet file with a weird name on Windows, maybe the encoding issue

**Comment by jnewbery** [comment]
Please update title to indicate that this PR is for GUI support for wallets loaded at runtime.

Please rebase on #10740 (which I just rebased on master) to pick up the *wallet: Make vpwallets usage thread safe* commit

**Comment by promag** [comment]
@jnewbery done.

**Comment by jnewbery** [comment]
Tested ACK c9048b9213f7e9dc9b8da9b06c2caccd27f8c3cc

**Comment by jnewbery** [comment]
This (and #13100) need to update doc/release-notes-pr10740.md to state that dynamic wallet loading *is* supported in the GUI.

**Comment by promag** [comment]
@jnewbery done.

**Comment by MeshCollider** [comment]
Concept ACK

**Comment by jnewbery** [comment]
@promag - this requires a rebase for one of my commits. I've rebased #10740. I suggest you reset to that branch and then cherrypick your commits.

**Comment by promag** [comment]
@jnewbery rebased.

**Comment by promag** [comment]
Rebased.

**Comment by mess110** [comment]
Tested ACK 0e674ba

**Comment by jnewbery** [comment]
Tested ACK 0e674ba5572275bcc1eb8f91c6292f871925f965

**Comment by jonasschnelli** [comment]
Started in regtest with only the "default" wallet loaded (single wallet mode), then, loaded another wallet. I could see the 2nd wallet, but the initial "default" wallet did had no name in the dropdown.

**Comment by jonasschnelli** [comment]
If you load a wallet when you are in the "Receive"-Tab (or any other tab), select the wallet via the Dropdown the "Overview"-view will appear instead of the "Receive"-View.

bildschirmfoto 2018-05-17 um 21 57 58

**Comment by jonasschnelli** [comment]
utACK 0e674ba5572275bcc1eb8f91c6292f871925f965 though I would prefer if the glitches would be fixed before merge

**Comment by jnewbery** [comment]
>  the initial "default" wallet did had no name in the dropdown.

Yes - this is a result of https://github.com/bitcoin/bitcoin/pull/11687#discussion_r158127240 , and is also visible in the `listwallets` and `getwalletinfo` RPC calls.

This PR does introduce that behaviour, but does make it more visible when loading a second wallet.

**Comment by promag** [comment]
@jonasschnelli launching with `bitcoin-qt -regtest -wallet -wallet=w1` also adds an empty item to the checkbox.

**Comment by promag** [comment]
@jonasnick care to test again? Will squash after that.

**Comment by jonasschnelli** [comment]
Tested ACK 2e751347190049ac1b7fb8378db004956700aaa1

**Comment by ken2812221** [comment]
default
default

If wallet failed to load because of pruned mode, the dropdown menu shows the wallet.

**Pull request merged**